Olá, Gustavo

Meu nome é Ramon. Ao ao longo do texto farei algumas observações sobre melhorias no código. Nas partes em que não houver comentários, está tudo certo, ok? Estarei aberto a feedbacks e discussões sobre o tema.

**Peço que mantenha e não altere os comentários que eu fizer por aqui para que possamos nos localizar posteriormente, ok?**

Mais uma coisa, vamos utilizar um código de cores para você entender os meus feedbacks no seu notebook. Funciona assim:

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Sucesso. Tudo foi feito corretamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Alerta não crítco, mas que pode ser corrigido para melhoria geral no seu código/análise
</div>

<div class="alert alert-block alert-danger">

<b>Comentário do revisor: </b> <a class="tocSkip"></a>
    
Erro que precisa ser consertado, caso contrário seu projeto **não** será aceito.
</div>

Você pode me responder usando isso:
<div class="alert alert-block alert-info">
<b>Resposta do Aluno.</b> <a class="tocSkip"></a>
</div>

###

<div class="alert alert-block alert-success">
<b>Comentário Geral do Revisor</b> <a class="tocSkip"></a>

Obrigado por enviar seu projeto. 
    
Parabéns, você fez um bom trabalho.
    
<span class="text-danger">Porém, há espaço para melhorias conforme os itens apontados anteriormente. Destaco os  pontos a seguir: </span>
 * <span class="text-danger"> Há alguns SQLs que precisam de ajuste. Deixei dicas. </span>


Portanto, sugiro que você corrija os pontos indicados.
    
Se você tiver alguma dúvida sobre de algo que você não tenha entendido, sinta-se à vontade para perguntar ao tutor ou deixar um comentário em **azul** para que, na próxima iteração, eu possa responder.

Conte conosco. Lembre-se de que você pode tirar dúvidas com revisores usando os comentários ou buscar ajuda de um tutor.
Bom trabalho. 


</div>




<div class="alert alert-block alert-success">
<b> Comentário final do revisor: </b> <a class="tocSkip"></a>

Tudo certo agora. Desejo sucesso na jornada. Parabéns pelo empenho.
</div>

# Objetivos do Estudo

O objetivo deste estudo é analisar o comportamento dos leitores a partir de um banco de dados contendo informações sobre livros, autores, editoras, classificações e resenhas de usuários. A análise é parte do processo de avaliação de mercado para o desenvolvimento de um novo aplicativo voltado a leitores, em resposta ao aumento do consumo de livros durante o período da pandemia.

As perguntas-chave do estudo incluem:

- Quais editoras e autores têm maior presença no mercado?
- Quais livros são mais bem avaliados e quais recebem mais atenção dos leitores?
- Como se comportam os leitores mais engajados?
- Qual o perfil das obras com maior volume e qualidade de avaliações?

Essas análises permitirão compreender melhor as preferências dos leitores e identificar padrões relevantes para apoiar decisões estratégicas no desenvolvimento de um novo produto digital para esse público.


In [2]:
# Bibliotecas
import pandas as pd
from sqlalchemy import create_engine

db_config = {
    'user': 'practicum_student',
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
    'port': 5432,
    'db': 'data-analyst-final-project-db'
}

connection_string = f"postgresql://{db_config['user']}:{db_config['pwd']}@{db_config['host']}:{db_config['port']}/{db_config['db']}"
engine = create_engine(connection_string, connect_args={'sslmode': 'require'})


### Exploração Inicial das Tabelas

Antes de iniciar as consultas analíticas, observei as primeiras linhas de cada tabela para entender sua estrutura e confirmar a consistência dos dados disponíveis.

As tabelas disponíveis são:

- `books`: contém informações sobre livros, incluindo título, número de páginas, data de publicação, autor e editora.
- `authors`: contém os nomes dos autores, relacionados via `author_id`.
- `publishers`: contém os nomes das editoras, relacionados via `publisher_id`.
- `ratings`: traz as avaliações numéricas feitas por usuários sobre os livros.
- `reviews`: contém as resenhas textuais feitas por usuários sobre os livros.

A visualização inicial ajudou a entender os relacionamentos entre as tabelas e definir a abordagem para as consultas SQL a seguir.


In [3]:
# livros
pd.read_sql('SELECT * FROM books LIMIT 5', con=engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
# autores
pd.read_sql('SELECT * FROM authors LIMIT 5', con=engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
# editoras
pd.read_sql('SELECT * FROM publishers LIMIT 5', con=engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [6]:
# avaliações numéricas
pd.read_sql('SELECT * FROM ratings LIMIT 5', con=engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
# resenhas textuais
pd.read_sql('SELECT * FROM reviews LIMIT 5', con=engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [8]:
# Encontre o número de livros lançados depois de 1 de janeiro de 2000

query = '''
SELECT COUNT(*) AS livros_pos_2000
FROM books
WHERE publication_date > '2000-01-01';
'''

pd.read_sql(query, con=engine)



livros_pos_2000
0              819

### Tarefa 1: Livros lançados após 1º de janeiro de 2000

Foi realizada uma consulta para contar quantos livros foram publicados após a data de 01/01/2000.

**Resultado:** 819 livros foram lançados após essa data.

```sql
SELECT COUNT(*) AS livros_pos_2000
FROM books
WHERE publication_date > '2000-01-01';


In [9]:
# Encontre o número de avaliações e a classificação média para cada livro

query = '''
SELECT
    b.title,
    r.book_id,
    COUNT(*) AS total_avaliacoes,
    ROUND(AVG(r.rating), 2) AS media_classificacao
FROM ratings AS r
JOIN books AS b ON r.book_id = b.book_id
GROUP BY r.book_id, b.title
ORDER BY total_avaliacoes DESC
LIMIT 10;
'''

pd.read_sql(query, con=engine)



title  book_id  \
0                            Twilight (Twilight  #1)      948   
1                The Hobbit  or There and Back Again      750   
2                             The Catcher in the Rye      673   
3               Angels & Demons (Robert Langdon  #1)       75   
4  Harry Potter and the Prisoner of Azkaban (Harr...      302   
5  Harry Potter and the Chamber of Secrets (Harry...      299   
6  Harry Potter and the Order of the Phoenix (Har...      301   
7                                        Animal Farm       79   
8  The Fellowship of the Ring (The Lord of the Ri...      722   
9  Harry Potter and the Half-Blood Prince (Harry ...      300   

   total_avaliacoes  media_classificacao  
0               160                 3.66  
1                88                 4.13  
2                86                 3.83  
3                84                 3.68  
4                82                 4.41  
5                80                 4.29  
6                75                 4.19  
7                74                 3.73  
8                74                 4.39  
9                73                 4.25


<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Precisa apresnetar o nome. Deixo dica abaixo
</div>

In [1]:
# query_t2 = """SELECT
#               books.title,
#               subquery.avg_rating,
#               subquery.review_count
#               FROM
#                   (SELECT
#                        reviews.book_id as book_id,
#                        COUNT (DISTINCT reviews.review_id) AS review_count,
#                        AVG (ratings.rating) AS avg_rating
#                    FROM
#                        reviews
#                        INNER JOIN ratings ON ratings.book_id = reviews.book_id
#                    GROUP BY
#                        reviews.book_id
#                    ORDER BY
#                        review_count DESC,
#                        avg_rating DESC) AS subquery
#                    INNER JOIN books ON subquery.book_id = books.book_id
#               ORDER BY
#                   review_count DESC,
#                   avg_rating DESC
#               LIMIT 10"""


<div class="alert alert-block alert-warning">
<b> Comentário do revisor v2: </b> <a class="tocSkip"></a>

Vou considerar a resposta, pois de fato você usou o número de avaliações (como notas em vez de reviews). 
</div>

### ### 🔍 Tarefa 2: Número de avaliações e média de classificação por livro

Esta consulta identificou os livros mais avaliados da base, agora incluindo os respectivos títulos (`title`) para facilitar a interpretação dos resultados.

**Top 10 livros com maior número de avaliações:**

| Título | Total de Avaliações | Média de Classificação |
|--------|---------------------|-------------------------|
| Twilight (Twilight #1) | 160 | 3.66 |
| The Hobbit or There and Back Again | 88 | 4.13 |
| The Catcher in the Rye | 86 | 3.83 |
| Angels & Demons (Robert Langdon #1) | 84 | 3.68 |
| Harry Potter and the Prisoner of Azkaban | 82 | 4.41 |
| ... | ... | ... |

Esses dados ajudam a entender quais títulos atraem maior participação do público e como estão sendo avaliados por leitores em grande volume.


In [7]:
# Identifique a editora que lançou o maior número de livros com mais de 50 páginas

query = '''
SELECT
    p.publisher,
    COUNT(b.book_id) AS total_livros
FROM books AS b
JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY total_livros DESC
LIMIT 1;
'''

pd.read_sql(query, con=engine)


publisher  total_livros
0  Penguin Books            42

### Tarefa 3: Editora com mais livros com mais de 50 páginas

Para garantir que a análise se concentre em livros substanciais (excluindo brochuras e publicações curtas), filtrei os livros com mais de 50 páginas e agrupei os dados por editora.

**Resultado:** A editora com mais livros acima de 50 páginas foi **Penguin Books**, com **42 publicações**.

Essa informação pode ser útil para segmentar melhor os dados em análises futuras.


In [8]:
# Identifique o autor com a média mais alta de classificação de livros: olhe apenas para livros com pelo menos 50 classificações.

query = '''
WITH livros_filtrados AS (
    SELECT
        book_id,
        COUNT(*) AS total_ratings,
        AVG(rating) AS media_rating
    FROM ratings
    GROUP BY book_id
    HAVING COUNT(*) >= 50
)

SELECT
    a.author,
    ROUND(AVG(lf.media_rating), 2) AS media_geral_autor
FROM livros_filtrados AS lf
JOIN books AS b ON lf.book_id = b.book_id
JOIN authors AS a ON b.author_id = a.author_id
GROUP BY a.author
ORDER BY media_geral_autor DESC
LIMIT 1;
'''

pd.read_sql(query, con=engine)


author  media_geral_autor
0  J.K. Rowling/Mary GrandPré               4.28

### Tarefa 4: Autor com a maior média de classificação (livros com 50+ avaliações)

Nesta etapa, o objetivo era identificar o autor cuja média de avaliações de livros fosse a mais alta, considerando apenas livros que receberam **pelo menos 50 classificações** — isso garante maior confiabilidade estatística.

**Resultado:**  
- Autor: **J.K. Rowling/Mary GrandPré**  
- Média geral de classificação: **4.28**

Essa métrica ajuda a destacar autores com alto desempenho entre leitores, com base em volume relevante de feedback.


In [10]:
# Encontre o número médio de avaliações entre usuários que classificaram mais do que 50 livros.

query = '''
SELECT
    AVG(subquery2.review_cnt) AS avg_review_cnt
FROM (
    SELECT
        COUNT(reviews.review_id) AS review_cnt,
        subquery1.username
    FROM (
        SELECT
            username,
            COUNT(rating_id) AS rating_cnt
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 50
    ) AS subquery1
    INNER JOIN reviews ON reviews.username = subquery1.username
    GROUP BY subquery1.username
) AS subquery2;
'''

pd.read_sql(query, con=engine)


avg_review_cnt
0       24.333333


<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Query está incorreta pois precisa usar reviews em vez de ratings. Dica abaixo:
</div>

In [2]:
# query_t5 = """SELECT
#                   AVG (subquery2.review_cnt) AS avg_review_cnt
#               FROM
#                   (SELECT
#                       COUNT (reviews.review_id) as review_cnt,
#                       subquery1.username
#                   FROM
#                       (SELECT
#                           username,
#                           COUNT (rating_id) AS rating_cnt
#                       FROM
#                           ratings
#                       GROUP BY
#                           username
#                       HAVING
#                           COUNT (rating_id) > 50) AS subquery1
#                       INNER JOIN reviews ON reviews.username = subquery1.username
#                   GROUP BY
#                       subquery1.username) AS subquery2"""

<div class="alert alert-block alert-success">
<b> Comentário do revisor v2: </b> <a class="tocSkip"></a>

correto

</div>

### Tarefa 5: Média de resenhas entre usuários que classificaram mais de 50 livros

O objetivo desta tarefa era calcular, entre os usuários que classificaram mais de 50 livros (`ratings`), qual o número médio de resenhas textuais (`reviews`) que eles escreveram.

**Resultado:**  
Esses usuários escreveram, em média, **24.33 resenhas**.

Esse dado mostra que o público mais engajado não apenas avalia muitos livros, como também costuma contribuir com feedbacks mais completos.  
Isso pode ser relevante para funcionalidades como sistema de reputação ou curadoria de conteúdo no novo aplicativo.



### Conclusão do Projeto SQL

A análise realizada com SQL permitiu extrair insights relevantes sobre o comportamento dos leitores e a qualidade do acervo, com base em livros, autores, editoras, classificações e resenhas disponíveis na base de dados.

Principais descobertas:

- O catálogo possui **819 livros lançados após o ano 2000**, indicando boa oferta de publicações contemporâneas.
- Os livros mais populares em número de avaliações incluem títulos conhecidos como *Twilight*, *The Hobbit* e obras da saga *Harry Potter*, com destaque tanto em quantidade de avaliações quanto em média de nota.
- A editora **Penguin Books** foi a que mais publicou livros com mais de 50 páginas, totalizando **42 títulos** — o que a destaca na categoria de publicações mais substanciais.
- O autor **J.K. Rowling/Mary GrandPré** apresentou a melhor média de classificação entre livros com pelo menos 50 avaliações, com nota média de **4.28**.
- Entre os leitores mais ativos (aqueles que classificaram mais de 50 livros), a média de **resenhas escritas** foi de **24.33**, revelando um grupo de usuários engajados não só em avaliar, mas também em compartilhar opiniões mais completas.

Esses achados ajudam a delinear oportunidades para o desenvolvimento de um aplicativo voltado a leitores, identificando conteúdos e perfis que merecem atenção especial na construção de funcionalidades e estratégias de recomendação.

